DOC Add a 'Cython Best Practices, Conventions and Knowledge' section (#25608)
scikit-learn/scikit-learn: doc/developers/cython.rst

  objects, which include functions). In this regard, `PEP073 <https://peps.python.org/pep-0703/>`_
  provides a good overview and context and pathways for removal.

* Make sure you have deactivated `checks <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/sklearn/_build_utils/__init__.py#L68-L87>`_.

* Always prefer memoryviews instead over ``cnp.ndarray`` when possible: memoryviews are lightweight.

* Avoid memoryview slicing: memoryview slicing might be costly or misleading in some cases and
  we better not use it, even if handling fewer dimensions in some context would be preferable.

* Decorate final classes or methods with ``@final`` (this allows removing virtual tables when needed)

* Inline methods and function when it makes sense

* Make sure your Cython compilation units `use NumPy recent C API <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/setup.py#L64-L70>`_.

* In doubt, read the generated C or C++ code if you can: "The fewer C instructions and indirections
  for a line of Cython code, the better" is a good rule of thumb.

> * Understand that ``nogil`` declarations are just hints: when declaring the ``cdef`` functions

Reviewer: ```suggestion
* ``nogil`` declarations are just hints: when declaring the ``cdef`` functions
```

  the computationally intensive sections of the algorithms.
  For more information, please refer to `this section of Cython's tutorial <https://cython.readthedocs.io/en/latest/src/tutorial/cython_tutorial.html#primes>`_

    .. code-block::

         # This generates a HTML report (`source.html`) for `source.c`.
         cythonX --annotate source.pyx

Tips for performance
^^^^^^^^^^^^^^^^^^^^

* Understand the GIL in context for CPython (which problems it solves, what are its limitations)
  and get a good understanding of when Cython will be mapped to C code free of interactions with
  CPython, when it will not, and when it cannot (e.g. presence of interactions with Python
  objects, which include functions). In this regard, `PEP073 <https://peps.python.org/pep-0703/>`_
  provides a good overview and context and pathways for removal.

* Make sure you have deactivated `checks <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/sklearn/_build_utils/__init__.py#L68-L87>`_.

* Always prefer memoryviews instead over ``cnp.ndarray`` when possible: memoryviews are lightweight.

> * Avoid memoryview slicing: memoryview slicing might be costly or misleading in some cases and

Reviewer 2: FWIW I have an PR that should dramatically improve this. It almost certainly won't (and shouldn't!) make it into Cython 3, but hopefully won't be long after that.  Essentially, a lot of the reference counting turns out to be pretty pointless since you're usually incrementing and decrementing exactly the same object so it can be skipped completely.

So hopefully this advice can change sometime in the next year. But not yet

Author: Thanks for mentioning that, @da-woods!

Author: Thank you for mentioning this, @da-woods.

  we better not use it, even if handling fewer dimensions in some context would be preferable.

* Decorate final classes or methods with ``@final`` (this allows removing virtual tables when needed)

* Inline methods and function when it makes sense

* Make sure your Cython compilation units `use NumPy recent C API <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/setup.py#L64-L70>`_.

* In doubt, read the generated C or C++ code if you can: "The fewer C instructions and indirections
  for a line of Cython code, the better" is a good rule of thumb.

* ``nogil`` declarations are just hints: when declaring the ``cdef`` functions
  as nogil, it means that they can be called without holding the GIL, but it does not release
  the GIL when entering them. You have to do that yourself either by passing ``nogil=True`` to
  ``cython.parallel.prange`` explicitly, or by using an explicit context manager:

.. code-block:: cython

   cdef inline void my_func(self) nogil:

        # Some logic interacting with CPython, e.g. allocating arrays via NumPy.

        with nogil:
            # The code here is run as is it were written in C.

        return 0

> This item is based on `this comment from Stéfan's Benhel <https://github.com/cython/cython/issues/2798#issuecomment-459971828>`_

Reviewer: Maybe this is the cause of the bad rendering. It should probably be indented

Author: Fixed by 7ffdae28775a9a23074c9dc04bd365b55877efe5.

Reviewer: it looks great now

* `Start from scratch in a notebook <https://cython.readthedocs.io/en/latest/src/quickstart/build.html#using-the-jupyter-notebook>`_ to understand how to use Cython and to get feedback on your work quickly.

* To debug C code (e.g. a segfault), do use ``gdb`` with:

  .. code-block:: bash

         gdb --ex r --args python ./entrypoint_to_bug_reproducer.py

* To have access to some value in place to debug in ``cdef (nogil)`` context, use:

  .. code-block:: cython

         with gil:
             print(state_to_print)

* Note that Cython cannot parse f-strings with ``{var=}`` expressions, e.g.

  .. code-block:: bash

         print(f"{test_val=}")

* scikit-learn codebase has a lot of non-unified (fused) types (re)definitions. There currently is `ongoing work to simplify
  and unify that across the codebase <https://github.com/scikit-learn/scikit-learn/issues/25572>`_. For now, make sure you understand which concrete types are used ultimately.

* It is helpful to use ``gdb`` to debug. In order to do so, one must use a Python interpreter built with debug support
  (debug symbols and proper optimization). To create a new conda environment (which you might need to deactivate and

>   reactivate conda after building/installing) with a source-built CPython interpreter:

Reviewer: I'm still not sure that this section belongs to cython tips

Author: I'll move it to another file.

         make EXTRA_CFLAGS='-DPy_DEBUG' -j<num_cores>
         make install

  * You might find this alias to compile individual Cython extension handy:

    .. code-block::

         # You might want to add this alias to your shell script config.
         alias cythonX="cython -X language_level=3 -X boundscheck=False -X wraparound=False -X initializedcheck=False -X nonecheck=False -X cdivision=True"

         # This generates `source.c` as as if you had recompiled scikit-learn entirely.
         cythonX --annotate source.pyx

  * Using the ``--annotate`` option with this flag allows generating a HTML report of code annotation.
    This report reports interactions with the CPython interpreter on a line-by-line basis.
    Interactions with the CPython interpreter must be avoided as much as possible.
    For more information, please refer to `this section of Cython's tutorial <https://cython.readthedocs.io/en/latest/src/tutorial/cython_tutorial.html#primes>`_

    .. code-block::

         # This generates a HTML report (`source.html`) for `source.c`.
         cythonX --annotate source.pyx

Tips for performance
^^^^^^^^^^^^^^^^^^^^

> 

Reviewer: No, I'm talking about the rendering.
![render](https://user-images.githubusercontent.com/34657725/223744900-a53beaaf-7940-4c70-ab10-b9798bdbe261.png)
The bullet points from the section "Tips for performance" are tight and a little bit more to the left compared to the bullet points of the section above.

> 

Author: Honestly, I do not not understand this difference in rendering.

* If you intend to use OpenMP: On MacOS, system's distribution of ``clang`` does not implement OpenMP.
  You can install the ``compilers`` package available on ``conda-forge`` which comes with implementations of OpenMP.

* Activating `checks <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/sklearn/_build_utils/__init__.py#L68-L87>`_ might help. E.g. for activating boundscheck use:

  .. code-block:: bash

         export SKLEARN_ENABLE_DEBUG_CYTHON_DIRECTIVES=1

* `Start from scratch in a notebook <https://cython.readthedocs.io/en/latest/src/quickstart/build.html#using-the-jupyter-notebook>`_ to understand how to use Cython and to get feedback on your work quickly.

* To debug C code (e.g. a segfault), do use ``gdb`` with:

  .. code-block:: bash

         gdb --ex r --args python ./entrypoint_to_bug_reproducer.py

* To have access to some value in place to debug in ``cdef (nogil)`` context, use:

  .. code-block:: cython

         with gil:
             print(state_to_print)

* Note that Cython cannot parse f-strings with ``{var=}`` expressions, e.g.

  .. code-block:: bash

         print(f"{test_val=}")

> * scikit-learn codebase is a bit of a zero of (fused) types (re)definitions, and there's `ongoing work to simplify

Author: A mistake. How does this rewording sound to you?

```suggestion
* scikit-learn codebase has a lot of non-unified (fused) types (re)definitions. There currently is `ongoing work to simplify
```

> * scikit-learn codebase is a bit of a zero of (fused) types (re)definitions, and there's `ongoing work to simplify

Reviewer 3: Ah okay! Makes sense now :)

  it cannot (e.g. presence of interactions with Python objects, which include functions). In this regard,
  `PEP073 <https://peps.python.org/pep-0703/>`_ provides a good overview and context and pathways for removal.

* Make sure you have deactivated `checks <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/sklearn/_build_utils/__init__.py#L68-L87>`_.

* Always prefer memoryviews instead over ``cnp.ndarray`` when possible: memoryview are lightweight.

* Avoid memoryview slicing: memoryview slicing might be costly or misleading in some cases and we better not use it IMO,
  even if handling fewer dimensions in some context would be preferable.

* Decorate final classes or methods with ``@final`` (this allows removing virtual tables when needed)

* Inline methods and function when it makes sense

* Make sure your Cython compilation units `use NumPy recent C API <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/setup.py#L64-L70>`_.

* In doubt, read the generated C or C++ code if you can: "The fewer C instructions and indirections for a line of
  Cython code, the better" is a good rule of thumb.

> Cython Internals

Reviewer 3: Is it possible to add some context to this section on why this might be relevant to devs of sklearn and possible best practices associated?

Author: This part better be introduced in Cython's documentation directly, yes.

----------------

Concrete method dispatch implementation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

* In most cases, concrete method dispatch is performed **at runtime using virtual tables**
  (i.e. static C-array of pointers to a given types of function, used at runtime to known which
  function to call from an instance). This is costly because there is a double indirection.

* In some cases, concrete method dispatch is performed **statically** (a function is called
  directly by a pointer dereference). This is less costly, especially if there are rooms
  for this function to be inline.

Known Limitations
^^^^^^^^^^^^^^^^^

* No template for types exist in Cython, alternatives are:

  * using fused-types

>   * using Tempita, a small preprocessing language to expand Cython source

Reviewer 3: Is it worth mentioning C-pointer casting as well? If it's a struct in the function signature. Or too inelegant?

https://www.codementor.io/@arpitbhayani/powering-inheritance-in-c-using-structure-composition-176sygr724

Author: I am not sure I understand.

How is it a limitation? Moreover, how would you phrase it?

Reviewer 3: Using fused-types is not possible if say the struct is a extension type attribute:

```Cython
cdef struct AStruct:
     double member

cdef class A:
    AStruct B

    cdef do_something(self):
          B.member = 0.0
```

So instead, if one wants to subclass A and add members to `B`, then one would need to either use tempita, or perform C-casting.

```Cython
cdef struct newAStruct:
     Astruct origstruct
     double newmember

cdef class newA(A):
    AStruct B

    cdef do_something(self):
          B = <newAStruct> self.B
          B.member = 0.0
          B.newmember = 0.0
```

Author: Have you found places where such a pattern has been used or would be useful?

Reviewer 3: I used it in the Oblique splitter PR #22754. Since it's not merged, not sure if it's warranted to put the idea onto the Cython doc. I proposed the thought to provide some extra Cython ideas. Feel free to veto it :p

Author: I guess we better add this then if such changes get into scikit-learn's codebase. Could you create a pull request in this case?

.. _cython:

Cython Best Practices, Conventions and Knowledge
================================================

This documents:

* technical aspects of Cython which aren't currently present in Cython's documentation
* tips to develop Cython code in scikit-learn

Cython Internals

> ----------------

Reviewer: This section presents advanced stuff. I think it should come after the tips which will be useful in more situations

  (i.e. static C-array of pointers to a given types of function, used at runtime to known which
  function to call from an instance). This is costly because there is a double indirection.

* In some cases, concrete method dispatch is performed **statically** (a function is called
  directly by a pointer dereference). This is less costly, especially if there are rooms
  for this function to be inline.

Known Limitations
^^^^^^^^^^^^^^^^^

* No template for types exist in Cython, alternatives are:

  * using fused-types
  * using Tempita, a small preprocessing language to expand Cython source

* Multiple inheritance is not possible for extension types
* No concept similar to interfaces or traits exist in Cython
* Currently it is impossible to inherit and override a method which has a fused type in its signature
* Structs cannot be created using constructor syntax with `nogil` `issue and workaround <https://github.com/cython/cython/issues/1642>`_

Performances improvements will come with `Cython 3.0 which is to be released soon in beta apparently <https://github.com/cython/cython/issues/4022#issuecomment-1340918590>`_

Basics
^^^^^^

> * ``nogil`` declaration: https://github.com/cython/cython/issues/2798#issuecomment-459971828

Reviewer: Instead of linking to a github comment, I'd rather rewrite the full explanation here. And probably explain in more detail in which situation we want to release the gil, i.e. when the code is meant to be used in a multi-threaded context.

Tips for developping Cython code in scikit-learn
------------------------------------------------

Tips to ease development
^^^^^^^^^^^^^^^^^^^^^^^^

* If you intend to use OpenMP: On MacOS, system's distribution of ``clang`` does not implement OpenMP.
  You can install the ``compilers`` package available on ``conda-forge`` which comes with implementations of OpenMP.

* Activating `checks <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/sklearn/_build_utils/__init__.py#L68-L87>`_ might help. E.g. for activating boundscheck use:

  .. code-block:: bash

         export SKLEARN_ENABLE_DEBUG_CYTHON_DIRECTIVES=1

> * Start from scratch in a notebook to understand.

Reviewer: You can give some details here:
- first load cython extension ``%load_ext cython``
- to run a cython cell ``%%cython``
- some useful flags
    - ``%%cython -f`` to force recompiling each time you run the cell
    - ``%%cython --compile-args=-fopenmp`` if trouble compiling openmp code (replace by the appropriate flag for the compiler)
- maybe even link to the cython magic doc https://ipython.org/ipython-doc/2/config/extensions/cythonmagic.html

Author: I think I better link to [this part of the documentation](https://cython.readthedocs.io/en/latest/src/quickstart/build.html#using-the-jupyter-notebook). Then we can contribute your proposed elements to Cython's documentation directly.

What do you think? :slightly_smiling_face:

Reviewer: I agree, better link to the official documentation.
I still think that it's worth mentioning ``--compile-args=-fopenmp`` if trouble compiling openmp code (replace by the appropriate flag for the compiler) because it's not in the linked page.

> * Start from scratch in a notebook to understand.
> * To debug C code (e.g. a segfault), do use ``gdb`` with:

Reviewer: nitpick
```suggestion
* Start from scratch in a notebook to understand.

* To debug C code (e.g. a segfault), do use ``gdb`` with:
```

  .. code-block:: bash

         gdb --ex r --args python ./entrypoint_to_bug_reproducer.py

* To have access to some value in place to debug in ``cdef (nogil)`` context, use:

  .. code-block:: cython

         with gil:
             print(state_to_print)

* Note that Cython cannot parse f-strings with ``{var=}`` expressions, e.g.

>     ``print(f"{test_val=}")``

Reviewer 2: This is in Cython 3 I believe. Although it is a current limitation for the stable releases

>     ``print(f"{test_val=}")``

Reviewer: I believe this is a note for the previous point but here it's displayed as a separate point

> * scikit-learn codebase is a bit of a zero of (fused) types (re)definitions, and we need to simplified
>   and unify that across the codebase. For now, make sure you understand which concrete types are used ultimately.

Reviewer: you can maybe add a link to the corresponding issue where the ongoing work is tracked https://github.com/scikit-learn/scikit-learn/issues/25572
```suggestion
* scikit-learn codebase is a bit of a zero of (fused) types (re)definitions, and there's ongoing work to simplify
  and unify that across the codebase. For now, make sure you understand which concrete types are used ultimately.
```

* It is helpful to use ``gdb`` to debug. In order to do so, one must use a Python interpreter built with debug support
  (debug symbols and proper optimization). To create a new conda environment with a source-built interpreter:

> 

Reviewer: Is it really linked to cython tips in particular ? Looks like a much more general statement

Author: I agree. I think @Micky774 proposed this change.

  .. code-block:: bash

         git clone https://github.com/python/cpython.git
         conda create -n debug-scikit-dev
         conda activate debug-scikit-dev
         cd cpython
         mkdir debug
         cd debug
         ../configure --prefix=$CONDA_PREFIX --with-pydebug
         make EXTRA_CFLAGS='-DPy_DEBUG' -j<num_cores>
         make install

> * You may need to restart conda after building/installing (I had to, but not sure why).

Reviewer: I believe this is a remark related to the previous point but it is rendered as a new point
```suggestion
* You may need to restart conda after building/installing.
```

  * You might find this alias handy:
    .. code-block::

>          alias cythonX="cython -X language_level=3 -X boundscheck=False -X wraparound=False -X initializedcheck=False -X nonecheck=False -X cdivision=True"

Reviewer: and this one is rendered as a sub item of the previous one while it should be a new one.
Also I think you can merge this with the point about checks.

  * which you can use with:

    .. code-block::

>          cythonX --annotate source.pyx

Reviewer: I think a separate point about ``--annotate`` is worth. You can explain what it does where to find the annotated code (in the ``.html``) and how to interpret it: the more yellow is the line the more interaction with the Python interpreter it generates; we want to avoid such interactions as much as possible in parts of the code that are called many times, e.g. inside loops.

Author: What you propose is valuable, indeed. Yet, I think it might be stated into Cython's documentation directly.

What do you think?

Reviewer 2: There's definitely some examples of annotation in the Cython documentation (https://cython.readthedocs.io/en/latest/src/tutorial/cython_tutorial.html#primes). Although I don't believe for a second we have everything right and in its best place so improvements are welcome.

Personally, I don't think it hurts if there's a bit of duplication of advice.

Tips for performances
^^^^^^^^^^^^^^^^^^^^^

> 

Reviewer: nitpick, please seprate the bullet points by a blank line for better readability

* Understand the GIL in context for CPython (which problems it solves, what are its limitations) and get a good
  understanding of when Cython will be mapped to C code free of interactions with CPython, when it will not, and when
  it cannot (e.g. presence of interactions with Python objects, which include functions).
  In this regard, `PEP073 <https://peps.python.org/pep-0703/>`_ provides a good overview and context and pathways for removal.
* Make sure you have deactivated `checks <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/sklearn/_build_utils/__init__.py#L68-L87>`_.
* Always prefer memoryviews instead over ``cnp.ndarray`` when possible: memoryview are lightweight.
* Avoid memoryview slicing: memoryview slicing might be costly or misleading in some cases and we better not use it IMO,

>   even if handling fewer dimensions in some context would be preferable.

Reviewer: About this point, in the cython guidelines you can find https://cython-guidelines.readthedocs.io/en/latest/articles/memoryviews_are_views_not_arrays.html#slicing-is-quick

Does it mean it's fast but there's still some overhead and it's not fast enough for us ?

Also you can link to the pinned issue https://github.com/scikit-learn/scikit-learn/issues/17299

Reviewer 2: There's definitely some overhead. It's relatively lightweight compared to (say) creating a new memoryview, and it can be done without the GIL. But it isn't free.

I assume this based on genuine hard-earned experience within Pandas.

There's also the general rule that "user code should aim to be clean and understandable, while library code might want to squeeze out every little bit of performance". I assume Pandas is more like library code in this case, so my recommendation would be a bit different than to the average user.

I'll try to add a bit more detail in the guidelines... But I think the recommendation here is reasonable

* Decorate final classes or methods with ``@final`` (this allows removing virtual tables when needed)
* Inline methods and function when it makes sense
* Make sure your Cython compilation units `use NumPy recent C API <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/setup.py#L64-L70>`_.

> * In doubt, read the generated C or C++ code if you can: "The fewer C instructions and indirections for a line of Cython code, the better" is a good rule of thumb.

Reviewer 2: This is a good point. Obviously the annotated html is kind of helpful, but is not substitute for actually digging through the C code. Worth noting that the C code is annotated with the relevant Cython code so you can often search for the line of Cython code in your text editor to get to the relevant bits of C.

Experimental features
^^^^^^^^^^^^^^^^^^^^^

> If a function is called a lot of time in a for-loop, it may be useful to inline it to alleviate the overhead and gain performances.

Reviewer: I don't understand this section. There's already a point about inlining and I don't see the connection with experimental features

Author: I do not know who wrote this part. I guess this better be part of Cython's documentation.
